packages/rum-recorder/src/boot/recorder.spec.ts

```diff
@@ -0,0 +1,136 @@
+      expect(requests).toEqual([{ data: jasmine.any(FormData), size: jasmine.any(Number) }])
+      expect(formDataAsObject(requests[0].data)).toEqual({
+        'application.id': 'appId',
+        creation_reason: 'init',
+        end: jasmine.stringMatching(/^\d{13}$/),
+        has_full_snapshot: 'true',
+        records_count: '2',
+        segment: jasmine.any(File),
+        'session.id': 'session-id',
+        start: jasmine.stringMatching(/^\d{13}$/),
+        'view.id': 'view-id',
+      })
+      done()
+    })
+
+    const { lifeCycle } = setupBuilder.build()
+    renewSegment(lifeCycle)
+  })
+
+  it('renews the segment when its compressed data is getting too large', (done) => {
+    const clickCount = 10_000
+    waitRequests((requests) => {
+      expect(requests.length).toBe(1)
+      expect(requests[0].data.get('records_count')).toBe(String(clickCount + 2))
+      done()
+    })
+
+    setupBuilder.build()
+    const click = createNewEvent('click')
+    for (let i = 0; i < clickCount; i += 1) {
+      document.body.dispatchEvent(click)
+    }
+  })
+
+  it('stops sending new segment when the session is expired', (done) => {
+    waitRequests((requests) => {
+      expect(requests.length).toBe(1)
+      expect(requests[0].data.get('records_count')).toBe('3')
+      done()
+    })
+
+    const { lifeCycle } = setupBuilder.build()
+
+    document.body.dispatchEvent(createNewEvent('click'))
+
+    sessionId = undefined
+    renewSegment(lifeCycle)
+    document.body.dispatchEvent(createNewEvent('click'))
+
+    renewSegment(lifeCycle)
+  })
+
+  it('restarts sending segments when the session is renewed', (done) => {
+    waitRequests((requests) => {
+      expect(requests.length).toBe(1)
+      expect(requests[0].data.get('records_count')).toBe('1')
+      expect(requests[0].data.get('session.id')).toBe('new-session-id')
+      done()
+    })
+
+    sessionId = undefined
+    const { lifeCycle } = setupBuilder.build()
+
+    document.body.dispatchEvent(createNewEvent('click'))
+
+    sessionId = 'new-session-id'
+    renewSegment(lifeCycle)
+    document.body.dispatchEvent(createNewEvent('click'))
+
+    renewSegment(lifeCycle)
+  })
+})
+
+function renewSegment(lifeCycle: LifeCycle) {
+  lifeCycle.notify(LifeCycleEventType.BEFORE_UNLOAD)
+}
+
+function formDataAsObject(data: FormData) {
+  const result: { [key: string]: unknown } = {}
+  data.forEach((value, key) => {
+    result[key] = value
+  })
+  return result
+}
+
+function waitRequests(callback: (requests: Array<{ data: FormData; size: number }>) => void) {
+  const requests: Array<{ data: FormData; size: number }> = []
+  // Throttle the callback, so it is called only after the last request being sent
+  const { throttled: throttledCallback } = throttle(() => callback(requests), 300, { leading: false })
+  spyOn(HttpRequest.prototype, 'send').and.callFake((data: FormData, size) => {
+    requests.push({ data, size })
+    throttledCallback()
+  })
+}
```

Comment:
`send` is called asynchronously because of postmessage communications with the worker. By using `throttle`, I ensure that all requests generated by the test are being take into account. If I set an expected amount of requests, and wait only for them, my test will succeed even if extra requests are being sent.

I could move the `waitRequests` calls at the end of each test. It should work because we know that `send` is always called asynchronously, but if this changes, tests will break. It may be fine though.

Reply:
oh ok. do we really need the throttle? could having a `waitPostMessages(fn)` method that performs a `setTimeout(fn, xxx)` be enough?